File: Samples/XamlNavigation/cpp/NavMenuListView.cpp
#include "pch.h"
#include "NavMenuListView.h"

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::System;
using namespace Windows::UI::Core;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Documents;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Media::Animation;

namespace NavigationMenuSample
{
    namespace Controls
    {
        NavMenuListView::NavMenuListView()
        {
            this->SelectionMode = ListViewSelectionMode::Single;
            this->SingleSelectionFollowsFocus = false;
            this->IsItemClickEnabled = true;
            this->ItemClick += ref new ItemClickEventHandler(this, &NavMenuListView::ItemClickHandler);

            // Locate the hosting SplitView control
            this->Loaded += ref new Windows::UI::Xaml::RoutedEventHandler(this, &NavigationMenuSample::Controls::NavMenuListView::OnLoaded);
        }

        void NavMenuListView::OnLoaded(Platform::Object ^sender, Windows::UI::Xaml::RoutedEventArgs ^e)
        {
            auto parent = VisualTreeHelper::GetParent(this);
            while (parent != nullptr && dynamic_cast<SplitView^>(parent) == nullptr)
            {
                parent = VisualTreeHelper::GetParent((DependencyObject^)parent);
            }

            if (parent != nullptr)
            {
                _splitViewHost = dynamic_cast<SplitView^>(parent);


                _splitViewHost->RegisterPropertyChangedCallback(
                    SplitView::IsPaneOpenProperty,
                    ref new DependencyPropertyChangedCallback(this, &NavMenuListView::IsOpenPanePropertyChangedCallback));

                _splitViewHost->RegisterPropertyChangedCallback(
                    SplitView::DisplayModeProperty,
                    ref new DependencyPropertyChangedCallback(this, &NavMenuListView::DisplayModePropertyChangedCallback));

                // Call once to ensure we're in the correct state
                OnPaneToggled();
            }
        }

        void NavMenuListView::IsOpenPanePropertyChangedCallback(DependencyObject^ sender, DependencyProperty^ args)
        {
            OnPaneToggled();
        }

        void NavMenuListView::DisplayModePropertyChangedCallback(DependencyObject^ sender, DependencyProperty^ args)
        {
            OnPaneToggled();
        }

        void NavMenuListView::OnApplyTemplate()
        {
            ListView::OnApplyTemplate();

            // Remove the entrance animation on the item containers.
            for (int i = 0; i < (int)ItemContainerTransitions->Size; i++)
            {
                if (dynamic_cast<EntranceThemeTransition^>(ItemContainerTransitions->GetAt(i)) != nullptr)
                {
                    ItemContainerTransitions->RemoveAt(i);
                }
            }
        }

        /// <summary>
        /// Mark the <paramref name="item"/> as selected and ensures everything else is not.
        /// If the <paramref name="item"/> is null then everything is unselected.
        /// </summary>
        /// <param name="item"></param>
        void NavMenuListView::SetSelectedItem(ListViewItem^ item)
        {
            int index = -1;
            if (item != nullptr)
            {
                index = IndexFromContainer(item);
            }

            for (int i = 0; i < (int)Items->Size; i++)
            {
                auto lvi = (ListViewItem^)ContainerFromIndex(i);
                if (i != index)
                {
                    lvi->IsSelected = false;
                }
                else if (i == index)
                {
                    lvi->IsSelected = true;
                }
            }
        }

	/// <summary>
        /// Custom keyboarding logic to enable movement via the arrow keys without triggering selection
        /// until a 'Space' or 'Enter' key is pressed.
        /// </summary>
        /// <param name="e"></param>
        void NavMenuListView::OnKeyDown(KeyRoutedEventArgs^ e)
        {
            auto focusedItem = FocusManager::GetFocusedElement();

            auto shiftKeyState = CoreWindow::GetForCurrentThread()->GetKeyState(VirtualKey::Shift);
            auto shiftKeyDown = (shiftKeyState & CoreVirtualKeyStates::Down) == CoreVirtualKeyStates::Down;

            switch (e->Key)
            {
            case VirtualKey::Up:
                this->TryMoveFocus(FocusNavigationDirection::Up);
                e->Handled = true;
                break;

            case VirtualKey::Down:
                this->TryMoveFocus(FocusNavigationDirection::Down);
                e->Handled = true;
                break;

            case VirtualKey::Space:
            case VirtualKey::Enter:
                // Fire our event using the item with current keyboard focus
                InvokeItem(focusedItem);
                e->Handled = true;
                break;

            default:
                ListView::OnKeyDown(e);
                break;
            }
        }

        /// <summary>
        /// This method is a work-around until the bug in FocusManager.TryMoveFocus is fixed.
        /// </summary>
        /// <param name="direction"></param>
        void NavMenuListView::TryMoveFocus(FocusNavigationDirection direction)
        {
            if (direction == FocusNavigationDirection::Next || direction == FocusNavigationDirection::Previous)
            {
                FocusManager::TryMoveFocus(direction);
            }
            else
            {
                auto control = dynamic_cast<Control^>(FocusManager::FindNextFocusableElement(direction));
                if (control != nullptr)
                {
                    control->Focus(Windows::UI::Xaml::FocusState::Programmatic);
                }
            }
        }

        void NavMenuListView::ItemClickHandler(Object^ sender, ItemClickEventArgs^ e)
        {
            // Triggered when the item is selected using something other than a keyboard
            auto item = ContainerFromItem(e->ClickedItem);
            InvokeItem(item);
        }

        void NavMenuListView::InvokeItem(Object^ focusedItem)
        {
            auto item = dynamic_cast<ListViewItem^>(focusedItem);
            SetSelectedItem(item);
            ItemInvoked(this, item);

            if (_splitViewHost->IsPaneOpen && (
                _splitViewHost->DisplayMode == SplitViewDisplayMode::CompactOverlay ||
                _splitViewHost->DisplayMode == SplitViewDisplayMode::Overlay))
            {
                _splitViewHost->IsPaneOpen = false;
            }
            if (item != nullptr)
            {
                item->Focus(Windows::UI::Xaml::FocusState::Programmatic);
            }
        }

        /// <summary>
        /// Re-size the ListView's Panel when the SplitView is compact so the items
        /// will fit within the visible space and correctly display a keyboard focus rect.
        /// </summary>
        void NavMenuListView::OnPaneToggled()
        {
            if (_splitViewHost->IsPaneOpen)
            {
                ItemsPanelRoot->ClearValue(FrameworkElement::WidthProperty);
                ItemsPanelRoot->ClearValue(FrameworkElement::HorizontalAlignmentProperty);
            }
            else if (_splitViewHost->DisplayMode == SplitViewDisplayMode::CompactInline ||
                _splitViewHost->DisplayMode == SplitViewDisplayMode::CompactOverlay)
            {
                ItemsPanelRoot->SetValue(FrameworkElement::WidthProperty, _splitViewHost->CompactPaneLength);
                ItemsPanelRoot->SetValue(FrameworkElement::HorizontalAlignmentProperty, Windows::UI::Xaml::HorizontalAlignment::Left);
            }
        }
    }
}

File: Samples/XamlNavigation/cs/Controls/NavMenuListView.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.System;
using Windows.UI.Core;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Media.Animation;

namespace NavigationMenuSample.Controls
{
    /// <summary>
    /// A specialized ListView to represent the items in the navigation menu.
    /// </summary>
    /// <remarks>
    /// This class handles the following:
    /// 1. Sizes the panel that hosts the items so they fit in the hosting pane.  Otherwise, the keyboard
    ///    may appear cut off on one side b/c the Pane clips instead of affecting layout.
    /// 2. Provides a single selection experience where keyboard focus can move without changing selection.
    ///    Both the 'Space' and 'Enter' keys will trigger selection.  The up/down arrow keys can move
    ///    keyboard focus without triggering selection.  This is different than the default behavior when
    ///    SelectionMode == Single.  The default behavior for a ListView in single selection requires using
    ///    the Ctrl + arrow key to move keyboard focus without triggering selection.  Users won't expect
    ///    this type of keyboarding model on the nav menu.
    /// </remarks>
    public class NavMenuListView : ListView
    {
        private SplitView splitViewHost;

        public NavMenuListView()
        {
            this.SelectionMode = ListViewSelectionMode.Single;
            this.SingleSelectionFollowsFocus = false;
            this.IsItemClickEnabled = true;
            this.ItemClick += ItemClickedHandler;

            // Locate the hosting SplitView control
            this.Loaded += (s, a) =>
            {
                var parent = VisualTreeHelper.GetParent(this);
                while (parent != null && !(parent is SplitView))
                {
                    parent = VisualTreeHelper.GetParent(parent);
                }

                if (parent != null)
                {
                    this.splitViewHost = parent as SplitView;

                    splitViewHost.RegisterPropertyChangedCallback(SplitView.IsPaneOpenProperty, (sender, args) =>
                    {
                        this.OnPaneToggled();
                    });

                    splitViewHost.RegisterPropertyChangedCallback(SplitView.DisplayModeProperty, (sender, args) =>
                    {
                        this.OnPaneToggled();
                    });

                    // Call once to ensure we're in the correct state
                    this.OnPaneToggled();
                }
            };
        }

        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            // Remove the entrance animation on the item containers.
            for (int i = 0; i < this.ItemContainerTransitions.Count; i++)
            {
                if (this.ItemContainerTransitions[i] is EntranceThemeTransition)
                {
                    this.ItemContainerTransitions.RemoveAt(i);
                }
            }
        }

        /// <summary>
        /// Mark the <paramref name="item"/> as selected and ensures everything else is not.
        /// If the <paramref name="item"/> is null then everything is unselected.
        /// </summary>
        /// <param name="item"></param>
        public void SetSelectedItem(ListViewItem item)
        {
            int index = -1;
            if (item != null)
            {
                index = this.IndexFromContainer(item);
            }

            for (int i = 0; i < this.Items.Count; i++)
            {
                var lvi = (ListViewItem)this.ContainerFromIndex(i);
                if (i != index)
                {
                    lvi.IsSelected = false;
                }
                else if (i == index)
                {
                    lvi.IsSelected = true;
                }
            }
        }

        /// <summary>
        /// Occurs when an item has been selected
        /// </summary>
        public event EventHandler<ListViewItem> ItemInvoked;

        /// <summary>
        /// Custom keyboarding logic to enable movement via the arrow keys without triggering selection 
        /// until a 'Space' or 'Enter' key is pressed. 
        /// </summary>
        /// <param name="e"></param>
        protected override void OnKeyDown(KeyRoutedEventArgs e)
        {
            var focusedItem = FocusManager.GetFocusedElement();

            switch (e.Key)
            {
                case VirtualKey.Up:
                    this.TryMoveFocus(FocusNavigationDirection.Up);
                    e.Handled = true;
                    break;

                case VirtualKey.Down:
                    this.TryMoveFocus(FocusNavigationDirection.Down);
                    e.Handled = true;
                    break;

                case VirtualKey.Space:
                case VirtualKey.Enter:
                    // Fire our event using the item with current keyboard focus
                    this.InvokeItem(focusedItem);
                    e.Handled = true;
                    break;

                default:
                    base.OnKeyDown(e);
                    break;
            }
        }

        /// <summary>
        /// This method is a work-around until the bug in FocusManager.TryMoveFocus is fixed.
        /// </summary>
        /// <param name="direction"></param>
        private void TryMoveFocus(FocusNavigationDirection direction)
        {
            if (direction == FocusNavigationDirection.Next || direction == FocusNavigationDirection.Previous)
            {
                FocusManager.TryMoveFocus(direction);
            }
            else
            {
                var control = FocusManager.FindNextFocusableElement(direction) as Control;
                if (control != null)
                {
                    control.Focus(FocusState.Programmatic);
                }
            }
        }

        private void ItemClickedHandler(object sender, ItemClickEventArgs e)
        {
            // Triggered when the item is selected using something other than a keyboard
            var item = this.ContainerFromItem(e.ClickedItem);
            this.InvokeItem(item);
        }

        private void InvokeItem(object focusedItem)
        {
            this.SetSelectedItem(focusedItem as ListViewItem);
            this.ItemInvoked?.Invoke(this, focusedItem as ListViewItem);

            if (this.splitViewHost.IsPaneOpen && (
                this.splitViewHost.DisplayMode == SplitViewDisplayMode.CompactOverlay ||
                this.splitViewHost.DisplayMode == SplitViewDisplayMode.Overlay))
            {
                this.splitViewHost.IsPaneOpen = false;
            }

            if (focusedItem is ListViewItem)
            {
                ((ListViewItem)focusedItem).Focus(FocusState.Programmatic);
            }
        }

        /// <summary>
        /// Re-size the ListView's Panel when the SplitView is compact so the items
        /// will fit within the visible space and correctly display a keyboard focus rect.
        /// </summary>
        private void OnPaneToggled()
        {
            if (this.splitViewHost.IsPaneOpen)
            {
                this.ItemsPanelRoot.ClearValue(FrameworkElement.WidthProperty);
                this.ItemsPanelRoot.ClearValue(FrameworkElement.HorizontalAlignmentProperty);
            }
            else if (this.splitViewHost.DisplayMode == SplitViewDisplayMode.CompactInline ||
                this.splitViewHost.DisplayMode == SplitViewDisplayMode.CompactOverlay)
            {
                this.ItemsPanelRoot.SetValue(FrameworkElement.WidthProperty, this.splitViewHost.CompactPaneLength);
                this.ItemsPanelRoot.SetValue(FrameworkElement.HorizontalAlignmentProperty, HorizontalAlignment.Left);
            }
        }
    }
}
